File: pekko-sample-distributed-data-scala/src/main/scala/sample/distributeddata/ReplicatedCache.scala
package sample.distributeddata

import org.apache.pekko.actor.typed.ActorRef
import org.apache.pekko.actor.typed.Behavior
import org.apache.pekko.actor.typed.scaladsl.Behaviors
import org.apache.pekko.cluster.ddata.LWWMap
import org.apache.pekko.cluster.ddata.LWWMapKey
import org.apache.pekko.cluster.ddata.Replicator._
import org.apache.pekko.cluster.ddata.SelfUniqueAddress
import org.apache.pekko.cluster.ddata.typed.scaladsl.DistributedData
import org.apache.pekko.cluster.ddata.typed.scaladsl.Replicator.{ Get, Update }

object ReplicatedCache {
  sealed trait Command
  final case class PutInCache(key: String, value: String) extends Command
  final case class GetFromCache(key: String, replyTo: ActorRef[Cached]) extends Command
  final case class Cached(key: String, value: Option[String])
  final case class Evict(key: String) extends Command
  private sealed trait InternalCommand extends Command
  private case class InternalGetResponse(key: String, replyTo: ActorRef[Cached],
      rsp: GetResponse[LWWMap[String, String]])
      extends InternalCommand
  private case class InternalUpdateResponse(rsp: UpdateResponse[LWWMap[String, String]]) extends InternalCommand

  def apply(): Behavior[Command] = Behaviors.setup { context =>
    DistributedData.withReplicatorMessageAdapter[Command, LWWMap[String, String]] { replicator =>
      implicit val node: SelfUniqueAddress = DistributedData(context.system).selfUniqueAddress

      def dataKey(entryKey: String): LWWMapKey[String, String] =
        LWWMapKey("cache-" + math.abs(entryKey.hashCode % 100))

      Behaviors.receiveMessage[Command] {
        case PutInCache(key, value) =>
          replicator.askUpdate(
            askReplyTo =>
              Update(dataKey(key), LWWMap.empty[String, String], WriteLocal, askReplyTo)(_ :+ (key -> value)),
            InternalUpdateResponse.apply)

          Behaviors.same

        case Evict(key) =>
          replicator.askUpdate(
            askReplyTo =>
              Update(dataKey(key), LWWMap.empty[String, String], WriteLocal, askReplyTo)(_.remove(node, key)),
            InternalUpdateResponse.apply)

          Behaviors.same

        case GetFromCache(key, replyTo) =>
          replicator.askGet(
            askReplyTo => Get(dataKey(key), ReadLocal, askReplyTo),
            rsp => InternalGetResponse(key, replyTo, rsp))

          Behaviors.same

        case InternalGetResponse(key, replyTo, g @ GetSuccess(_, _)) =>
          replyTo ! Cached(key, g.dataValue.get(key))
          Behaviors.same

        case InternalGetResponse(key, replyTo, _: NotFound[_]) =>
          replyTo ! Cached(key, None)
          Behaviors.same

        case _: InternalGetResponse    => Behaviors.same // ok
        case _: InternalUpdateResponse => Behaviors.same // ok
      }
    }
  }
}

File: pekko-sample-distributed-data-java/src/main/java/sample/distributeddata/ReplicatedCache.java
package sample.distributeddata;

import static org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.readLocal;
import static org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.writeLocal;

import java.util.Optional;
import scala.Option;

import org.apache.pekko.actor.typed.ActorRef;
import org.apache.pekko.actor.typed.Behavior;
import org.apache.pekko.actor.typed.javadsl.ActorContext;
import org.apache.pekko.actor.typed.javadsl.Behaviors;
import org.apache.pekko.cluster.ddata.Key;
import org.apache.pekko.cluster.ddata.LWWMap;
import org.apache.pekko.cluster.ddata.LWWMapKey;
import org.apache.pekko.cluster.ddata.SelfUniqueAddress;
import org.apache.pekko.cluster.ddata.typed.javadsl.DistributedData;
import org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.Get;
import org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.GetResponse;
import org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.GetSuccess;
import org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.NotFound;
import org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.Update;
import org.apache.pekko.cluster.ddata.typed.javadsl.Replicator.UpdateResponse;
import org.apache.pekko.cluster.ddata.typed.javadsl.ReplicatorMessageAdapter;

public class ReplicatedCache {

  public interface Command {}

  public record PutInCache(String key, String value) implements Command {}

  public record GetFromCache(String key, ActorRef<Cached> replyTo) implements Command {}

  public record Cached(String key, Optional<String> value) {
    @Override
    public String toString() {
      return "Cached [key=" + key + ", value=" + value + "]";
    }
  }

  public record Evict(String key) implements Command {}

  private interface InternalCommand extends Command {}

  private record InternalGetResponse(String key, ActorRef<Cached> replyTo, GetResponse<LWWMap<String, String>> rsp)
      implements InternalCommand {}

  private record InternalUpdateResponse(UpdateResponse<LWWMap<String, String>> rsp)
      implements InternalCommand {}

  public static Behavior<Command> create() {
    return Behaviors.setup(context ->
        DistributedData.withReplicatorMessageAdapter(
            (ReplicatorMessageAdapter<Command, LWWMap<String, String>> replicator) ->
                new ReplicatedCache(context, replicator).createBehavior()));
  }

  private final ReplicatorMessageAdapter<Command, LWWMap<String, String>> replicator;
  private final SelfUniqueAddress node;

  public ReplicatedCache(
      ActorContext<Command> context,
      ReplicatorMessageAdapter<Command, LWWMap<String, String>> replicator
  ) {
    this.replicator = replicator;
    node = DistributedData.get(context.getSystem()).selfUniqueAddress();
  }

  public Behavior<Command> createBehavior() {
    return Behaviors
        .receive(Command.class)
        .onMessage(PutInCache.class, cmd -> receivePutInCache(cmd.key(), cmd.value()))
        .onMessage(Evict.class, cmd -> receiveEvict(cmd.key()))
        .onMessage(GetFromCache.class, cmd -> receiveGetFromCache(cmd.key(), cmd.replyTo()))
        .onMessage(InternalGetResponse.class, this::onInternalGetResponse)
        .onMessage(InternalUpdateResponse.class, notUsed -> Behaviors.same())
        .build();
  }

  private Behavior<Command> receivePutInCache(String key, String value) {
    replicator.askUpdate(
        askReplyTo ->
            new Update<>(
                dataKey(key),
                LWWMap.empty(),
                writeLocal(),
                askReplyTo,
                curr -> curr.put(node, key, value)
            ),
        InternalUpdateResponse::new);

    return Behaviors.same();
  }

  private Behavior<Command> receiveEvict(String key) {
    replicator.askUpdate(
        askReplyTo ->
            new Update<>(
                dataKey(key),
                LWWMap.empty(),
                writeLocal(),
                askReplyTo,
                curr -> curr.remove(node, key)
            ),
        InternalUpdateResponse::new);

    return Behaviors.same();
  }

  private Behavior<Command> receiveGetFromCache(String key, ActorRef<Cached> replyTo) {
    replicator.askGet(
        askReplyTo -> new Get<>(dataKey(key), readLocal(), askReplyTo),
        rsp -> new InternalGetResponse(key, replyTo, rsp));

    return Behaviors.same();
  }

  private Behavior<Command> onInternalGetResponse(InternalGetResponse msg) {
    if (msg.rsp() instanceof GetSuccess<LWWMap<String, String>> success) {
      Option<String> valueOption = success.get(dataKey(msg.key())).get(msg.key());
      Optional<String> valueOptional = Optional.ofNullable(valueOption.isDefined() ? valueOption.get() : null);
      msg.replyTo().tell(new Cached(msg.key(), valueOptional));
    } else if (msg.rsp() instanceof NotFound) {
      msg.replyTo().tell(new Cached(msg.key(), Optional.empty()));
    }
    return Behaviors.same();
  }

  private Key<LWWMap<String, String>> dataKey(String entryKey) {
    return LWWMapKey.create("cache-" + Math.abs(entryKey.hashCode() % 100));
  }

}
